common/src/graph.rs

```diff
@@ -315,27 +290,58 @@ impl<N> Node<N> {
     }
 }
 
+// Iterates through the graph with breadth-first search
 pub struct TreeIter<'a, N> {
-    id: NodeId,
-    remaining: &'a [Node<N>],
+    queue: VecDeque<NodeId>,
+    visited: FxHashSet<NodeId>,
+    nodes: &'a FxHashMap<NodeId, Node<N>>,
+}
+
+impl<'a, N> TreeIter<'a, N> {
+    fn new(graph: &'a Graph<N>) -> Self {
+        let mut result = TreeIter {
+            queue: VecDeque::from([NodeId::ROOT]),
+            visited: FxHashSet::from_iter([NodeId::ROOT]),
+            nodes: &graph.nodes,
+        };
+
+        // Skip the root node
+        let _ = result.next_node();
+
+        result
+    }
+
+    // Returns the next Node in the traversal. The iterator returns its parent and parent side.
+    fn next_node(&mut self) -> Option<&Node<N>> {
+        let node_id = self.queue.pop_front()?;
+        let node = &self.nodes[&node_id];
+        for side in Side::iter() {
+            if let Some(neighbor) = node.neighbors[side as usize] {
+                if !self.visited.contains(&neighbor) {
+                    self.queue.push_back(neighbor);
+                    self.visited.insert(neighbor);
+                }
+            }
+        }
+        Some(node)
+    }
 }
 
 impl<N> Iterator for TreeIter<'_, N> {
     type Item = (Side, NodeId);
 
     fn next(&mut self) -> Option<Self::Item> {
-        let (node, rest) = self.remaining.split_first()?;
-        self.remaining = rest;
-        self.id = NodeId::from_idx(self.id.idx() + 1);
```

Comment:
Was `self.id` dead all along? Geez.

Reply:
Oh, so it is. I didn't realize that while making this PR (since the implementation had to change quite a bit), but it does look like the only read of `self.id` is used to set `self.id`.